[macOS / M1] cpu_percent() gives incorrect information

## Summary

* OS: macOS 11.4
* Architecture: ARM
* Psutil version: 5.8.0
* Python version: 3.9.4
* Type: core

## Description

Process.cpu_percent() always reports very low cpu usage, in contradiction to top and Activity monitor. This test program:

```
from concurrent.futures import ThreadPoolExecutor
import psutil
import time
from math import log

def work(n):
    m = 0
    for i in range(1, n**2):
        m += log(i)
    return True
    
executor = ThreadPoolExecutor(max_workers=2)
task = executor.submit(work, 10000)

proc = psutil.Process()
while not task.done():
    print(proc.cpu_percent(interval=1))
```

consistently reports 2.4% usage on the Mac (vs. ~100% usage, as expected, on linux with the same version of psutil).

@rmalouf looks like there are compilation issues:

```
2021-10-02T09:27:47.8665360Z   psutil/_psutil_osx.c:224:25: error: use of undeclared identifier 'info'; did you mean 'sinf'?
2021-10-02T09:27:47.8666080Z       mach_timebase_info(&info);
2021-10-02T09:27:47.8666510Z                           ^~~~
2021-10-02T09:27:47.8666870Z                           sinf
2021-10-02T09:27:47.8668040Z   /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/math.h:328:14: note: 'sinf' declared here
2021-10-02T09:27:47.8668860Z   extern float sinf(float);
2021-10-02T09:27:47.8669250Z                ^
2021-10-02T09:27:47.8670190Z   psutil/_psutil_osx.c:225:39: error: use of undeclared identifier 'info'
2021-10-02T09:27:47.8670860Z       total_user = pti.pti_total_user * info.numer;
2021-10-02T09:27:47.8671370Z                                         ^
2021-10-02T09:27:47.8672280Z   psutil/_psutil_osx.c:226:19: error: use of undeclared identifier 'info'
2021-10-02T09:27:47.8672930Z       total_user /= info.denom;
2021-10-02T09:27:47.8673330Z                     ^
2021-10-02T09:27:47.8674210Z   psutil/_psutil_osx.c:227:43: error: use of undeclared identifier 'info'
2021-10-02T09:27:47.8675150Z       total_system = pti.pti_total_system * info.numer;
2021-10-02T09:27:47.8675680Z                                             ^
2021-10-02T09:27:47.8676670Z   psutil/_psutil_osx.c:228:21: error: use of undeclared identifier 'info'
2021-10-02T09:27:47.8677690Z       total_system /= info.denom;
2021-10-02T09:27:47.8678100Z                       ^
2021-10-02T09:27:47.8678490Z   5 errors generated.
2021-10-02T09:27:47.8679390Z   error: command 'gcc' failed with exit status 1
```

--- a/psutil/_psutil_common.c
+++ b/psutil/_psutil_common.c
@@ -170,6 +170,15 @@ convert_kvm_err(const char *syscall, char *errbuf) {
 }
 #endif
 
+// ====================================================================
+// --- macOS
+// ====================================================================
+
+#ifdef PSUTIL_OSX
+#include <mach/mach_time.h>
+
+struct mach_timebase_info MACH_TIMEBASE_INFO;
+#endif
 
 // ====================================================================
 // --- Windows
@@ -403,6 +412,10 @@ psutil_setup(void) {
         return 1;
     GetSystemInfo(&PSUTIL_SYSTEM_INFO);
     InitializeCriticalSection(&PSUTIL_CRITICAL_SECTION);
+#endif
+
+#ifdef PSUTIL_OSX
+    mach_timebase_info(&MACH_TIMEBASE_INFO);
 #endif
 
     return 0;

--- a/psutil/_psutil_common.c
+++ b/psutil/_psutil_common.c
@@ -177,7 +177,7 @@ convert_kvm_err(const char *syscall, char *errbuf) {
 #ifdef PSUTIL_OSX
 #include <mach/mach_time.h>
 
-struct mach_timebase_info MACH_TIMEBASE_INFO;
+struct mach_timebase_info PSUTIL_MACH_TIMEBASE_INFO;
 #endif
 
 // ====================================================================
@@ -415,7 +415,7 @@ psutil_setup(void) {
 #endif
 
 #ifdef PSUTIL_OSX
-    mach_timebase_info(&MACH_TIMEBASE_INFO);
+    mach_timebase_info(&PSUTIL_MACH_TIMEBASE_INFO);
 #endif
 
     return 0;

--- a/psutil/_psutil_common.h
+++ b/psutil/_psutil_common.h
@@ -123,9 +123,9 @@ void convert_kvm_err(const char *syscall, char *errbuf);
 // ====================================================================
 
 #ifdef PSUTIL_OSX
-#include <mach/mach_time.h>
+    #include <mach/mach_time.h>
 
-extern struct mach_timebase_info MACH_TIMEBASE_INFO;
+    extern struct mach_timebase_info PSUTIL_MACH_TIMEBASE_INFO;
 #endif
 
 // ====================================================================

--- a/psutil/_psutil_common.h
+++ b/psutil/_psutil_common.h
@@ -119,6 +119,16 @@ int psutil_setup(void);
 void convert_kvm_err(const char *syscall, char *errbuf);
 
 // ====================================================================
+// --- macOS
+// ====================================================================
+
+#ifdef PSUTIL_OSX
+#include <mach/mach_time.h>
+
+extern struct mach_timebase_info MACH_TIMEBASE_INFO;
+#endif
+
+// ====================================================================
 // --- Windows
 // ====================================================================
 

--- a/psutil/_psutil_osx.c
+++ b/psutil/_psutil_osx.c
@@ -279,17 +279,18 @@ static PyObject *
 psutil_proc_pidtaskinfo_oneshot(PyObject *self, PyObject *args) {
     pid_t pid;
     struct proc_taskinfo pti;
-    uint64_t total_user, total_system;
+    uint64_t total_user;
+    uint64_t total_system;
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))
         return NULL;
     if (psutil_proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &pti, sizeof(pti)) <= 0)
         return NULL;
 
-    total_user = pti.pti_total_user * MACH_TIMEBASE_INFO.numer;
-    total_user /= MACH_TIMEBASE_INFO.denom;
-    total_system = pti.pti_total_system * MACH_TIMEBASE_INFO.numer;
-    total_system /= MACH_TIMEBASE_INFO.denom;
+    total_user = pti.pti_total_user * PSUTIL_MACH_TIMEBASE_INFO.numer;
+    total_user /= PSUTIL_MACH_TIMEBASE_INFO.denom;
+    total_system = pti.pti_total_system * PSUTIL_MACH_TIMEBASE_INFO.numer;
+    total_system /= PSUTIL_MACH_TIMEBASE_INFO.denom;
 
     return Py_BuildValue(
         "(ddKKkkkk)",

--- a/psutil/_psutil_osx.c
+++ b/psutil/_psutil_osx.c
@@ -22,7 +22,6 @@
 #include <pwd.h>
 #include <unistd.h>
 #include <mach/mach.h>
-#include <mach/mach_time.h>
 #include <mach/mach_vm.h>
 #include <mach/shared_region.h>
 
@@ -280,7 +279,6 @@ static PyObject *
 psutil_proc_pidtaskinfo_oneshot(PyObject *self, PyObject *args) {
     pid_t pid;
     struct proc_taskinfo pti;
-    struct mach_timebase_info info;
     uint64_t total_user, total_system;
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))
@@ -288,11 +286,10 @@ psutil_proc_pidtaskinfo_oneshot(PyObject *self, PyObject *args) {
     if (psutil_proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &pti, sizeof(pti)) <= 0)
         return NULL;
 
-    mach_timebase_info(&info);
-    total_user = pti.pti_total_user * info.numer;
-    total_user /= info.denom;
-    total_system = pti.pti_total_system * info.numer;
-    total_system /= info.denom;
+    total_user = pti.pti_total_user * MACH_TIMEBASE_INFO.numer;
+    total_user /= MACH_TIMEBASE_INFO.denom;
+    total_system = pti.pti_total_system * MACH_TIMEBASE_INFO.numer;
+    total_system /= MACH_TIMEBASE_INFO.denom;
 
     return Py_BuildValue(
         "(ddKKkkkk)",

--- a/psutil/_psutil_osx.c
+++ b/psutil/_psutil_osx.c
@@ -280,6 +280,7 @@ static PyObject *
 psutil_proc_pidtaskinfo_oneshot(PyObject *self, PyObject *args) {
     pid_t pid;
     struct proc_taskinfo pti;
+    struct mach_timebase_info info;
     uint64_t total_user, total_system;
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))

--- a/psutil/_psutil_osx.c
+++ b/psutil/_psutil_osx.c
@@ -22,6 +22,7 @@
 #include <pwd.h>
 #include <unistd.h>
 #include <mach/mach.h>
+#include <mach/mach_time.h>
 #include <mach/mach_vm.h>
 #include <mach/shared_region.h>
 
@@ -279,16 +280,23 @@ static PyObject *
 psutil_proc_pidtaskinfo_oneshot(PyObject *self, PyObject *args) {
     pid_t pid;
     struct proc_taskinfo pti;
+    uint64_t total_user, total_system;
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))
         return NULL;
     if (psutil_proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &pti, sizeof(pti)) <= 0)
         return NULL;
 
+    mach_timebase_info(&info);
+    total_user = pti.pti_total_user * info.numer;
+    total_user /= info.denom;
+    total_system = pti.pti_total_system * info.numer;
+    total_system /= info.denom;
+
     return Py_BuildValue(
         "(ddKKkkkk)",
-        (float)pti.pti_total_user / 1000000000.0,     // (float) cpu user time
-        (float)pti.pti_total_system / 1000000000.0,   // (float) cpu sys time
+        (float)total_user / 1000000000.0,     // (float) cpu user time
+        (float)total_system / 1000000000.0,   // (float) cpu sys time
         // Note about memory: determining other mem stats on macOS is a mess:
         // http://www.opensource.apple.com/source/top/top-67/libtop.c?txt
         // I just give up.

